src/lib/device_tree/src/lib.rs

```diff
@@ -148,6 +192,7 @@ impl<'a> Iterator for FdtIterator<'a> {
                     self.len_buf[self.depth] = cursor_string(&self.dt.struct_block, &mut self.cursor, &mut self.path_buf[self.depth]).unwrap() as usize;
                     self.cursor = align4(self.cursor);
                     self.depth += 1;
+                    // Note: This performs a copy of the whole path_buf array!
```

Comment:
Good eye! Ideally, we would return a reference to `self.path_buf` to avoid the copy.

The problem is the function signature is:

```
fn next(&mut self) -> Option<Entry<'a>> {
```

and not:

```
fn next(&'a mut self) -> Option<Entry<'a>> {
```

Without the extra `'a`, the the Rust borrow checker will complain about the lifetimes if we returned a reference to `self`.
We cannot chain the signature because we must implement `Iterator`.

Many people have told me we can solve this problem using "streaming iterators".

Reply:
Yes, I started looking at it and indeed it is not a trivial fix. I have some ideas I want to try, but before I start recommending bigger changes I wanted to familiarize myself with the codebase first.

Just as FYI what I want to try (outside of this PR):

- First is something similar to streaming iterators you mentioned. I am not sure if there is any code that uses this FdtIterator and takes some big advantage of the fact that this is a standard iterator and not some custom set of methods. If not, maybe we can just design better API for our usecase. With the new API, I would like to ensure that errors are also properly handled in it (as mentioned in one of the TODOs in this file, it's unclear what to do with errors in the iterators). I need a little more time and experimentation to propose something more concrete here.

- The other part I want to explore is whether we actually need the Path to contain the whole path, or could we just return the last component of the path (node name or property name), and let the user of the library properly handle reconstructing the whole path if they need to. This would have few interesting effects:

  - The copy would not be such a big problem, as we would only copy the buffer for one name (64 bytes now), instead of the whole 16*64 array of arrays.
  - This might remove the restriction on maximum nesting of the library (as we would never store the whole path, just the node we are currently in)
  - The disadvantage though is that if someone needs all components of the path, then such API might be harder to use. That is why I need to explore a little more how this library is used before proposing this change.

Reply:
The main usecase is iterating over the FDT and pulling out some specific values and storing them in variables on the stack.  It's for configuration at runtime. The tricky part is there is only stack memory, so no heap memory.

Feel free to re-design this. You can see the examples of how the API is used right now.

Also, take a look at how libfdt is designed for some ideas: https://github.com/dgibson/dtc/blob/master/libfdt/libfdt.h#L180